SMPyBandits/Policies/FEWA.py

```diff
@@ -0,0 +1,109 @@
+
+
+class EFF_FEWA(BasePolicy):
+    """
+    Efficient mechanism as described in [Seznec et al.,  2019a] (m=2) and [Seznec et al.,  2019b] (m<=2)
+    Algorithm Filtering on Expanding Window Average [Seznec et al.,  2019a]
+    We use the confidence level \delta_t = \frac{1}{t^\alpha}.
+     """
+    def __init__(self, nbArms, alpha=0.06, subgaussian=1, m=2):
+        super(EFF_FEWA, self).__init__(nbArms)
+        self.alpha = alpha
+        self.nbArms = nbArms
+        self.subgaussian = subgaussian
+        self.statistics = np.ones(shape=(3, self.nbArms, 2)) * np.nan
+        # [0,:,:] : current statistics, [1,:,:]: pending statistics, [3,:,:]: number of sample in the pending statistics
+        self.windows = np.array([1, int(np.ceil(m))])
+        self.outlogconst = np.sqrt(self.windows * sqrt(8 * self.alpha * self.subgaussian ** 2))
+        self.armSet = np.arange(nbArms)
+        self.grid = m
+
+    def __str__(self):
+        return r"EFF_FEWA($\alpha={:.3g}, \, m={:.3g}$)".format(self.alpha, self.grid)
+
+    def getReward(self, arm, reward):
+        super(EFF_FEWA, self).getReward(arm, reward)
+        if not np.all(np.isnan(self.statistics[0,:,-1])):
+            self.statistics = np.append(self.statistics, np.nan * np.ones([3, self.nbArms, 1]), axis=2)
+        while self.statistics.shape[2] > min([len(self.outlogconst), len(self.windows)]):
+            self.windows = np.append(self.windows, int(np.ceil(self.windows[-1] * self.grid)))
+            self.outlogconst = np.append(self.outlogconst, self._append_thresholds(self.windows[-1]))
+        self.statistics[1, arm, 0] = reward
+        self.statistics[2, arm, 0] = 1
+        self.statistics[1, arm, 1:] += reward
+        self.statistics[2, arm, 1:] += 1
+        idx = np.where((self.statistics[2, arm, :] == self.windows))[0]
+        self.statistics[0, arm, idx] = self.statistics[1, arm, idx]
+        idx_nan = np.where(np.isnan(self.statistics[2, arm, :]))[0]
+        idx = np.concatenate([idx, np.array([i for i in idx_nan if i - 1 in set(list(idx))]).astype(int)])
+        self.statistics[1:, arm, idx[idx != 0]] = self.statistics[1:, arm, idx[idx != 0] - 1]
+
+    def choice(self):
+        remainingArms = self.armSet.copy()
+        i = 0
+        selected = remainingArms[np.isnan(self.statistics[0, :, i])]
+        delta_inv = self._confidence_level_inv()
+        sqrtlogt = np.sqrt(np.log(delta_inv))
+        while len(selected) == 0 :
+            thresh = np.max(self.statistics[0, remainingArms, i]) - sqrtlogt * self.outlogconst[i]
+            remainingArms = remainingArms[self.statistics[0, remainingArms, i] >= thresh]
+            i += 1
+            selected = remainingArms[np.isnan(self.statistics[0, remainingArms, i])] if len(
+                remainingArms) != 1 else remainingArms
+        return selected[np.argmin(self.pulls[selected])]
+
+    def _append_thresholds(self, w):
+        return sqrt(8 * self.subgaussian ** 2 * w)
```

Comment:
I would write `8 * w * self.subgaussian**2` to make sure that `8 * w` are multiplied together, and not `self.subgaussian ** (2 * w)`.